File: Problema1/solucion.py
"""
UNIVERSIDAD NACIONAL EXPERIMENTAL DE GUAYANA
Asignatura: Lenguajes y Compiladores
Problema 1: Polinomio (x+1)^n usando Triángulo de Pascal

Explicación:
Este código implementa la generación de coeficientes binomiales mediante
el Triángulo de Pascal. Utiliza memoria dinámica (listas de Python) y
precisión arbitraria para manejar n=100.
"""
import time

def generar_fila_pascal(n):
    # Genera la fila n del Triángulo de Pascal
    fila = [1]
    for _ in range(n):
        nueva_fila = [1]
        for i in range(len(fila) - 1):
            nueva_fila.append(fila[i] + fila[i+1])
        nueva_fila.append(1)
        fila = nueva_fila
    return fila

def obtener_terminos_polinomio(coeficientes):
    n = len(coeficientes) - 1
    terminos = []
    for i, coeff in enumerate(coeficientes):
        potencia = n - i
        termino = ""
        if coeff != 1 or potencia == 0:
            termino += str(coeff)
        if potencia > 0:
            termino += "x"
            if potencia > 1:
                termino += f"^{potencia}"
        terminos.append(termino)
    return terminos

def calcular_paso_a_paso(x, n, coeficientes):
    resultado = 0
    pasos = []
    for i, coeff in enumerate(coeficientes):
        potencia = n - i
        valor_termino = coeff * (x ** potencia)
        resultado += valor_termino
        pasos.append(f"{coeff}*({x}^{potencia}) = {valor_termino}")
    return pasos, resultado

def ejecutar_problema1():
    print("--- CONFIGURACIÓN DEL POLINOMIO ---")
    
    # Entrada para n
    while True:
        entrada_n = input("Ingrese el valor de n (Enter para 100): ").strip()
        if not entrada_n:
            n = 100
            break
        try:
            n = int(entrada_n)
            if n >= 0: break
            print("Error: n debe ser un número no negativo.")
        except ValueError:
            print("Error: Entrada inválida. Ingrese un número entero.")

    # Entrada para x
    while True:
        entrada_x = input("Ingrese el valor de x para evaluar f(x) (Enter para 2): ").strip()
        if not entrada_x:
            x = 2
            break
        try:
            x = int(entrada_x)
            break
        except ValueError:
            print("Error: Entrada inválida. Ingrese un número entero.")
    
    # Benchmarking
    inicio = time.perf_counter()
    coeficientes = generar_fila_pascal(n)
    fin = time.perf_counter()
    
    terminos = obtener_terminos_polinomio(coeficientes)
    pasos, resultado = calcular_paso_a_paso(x, n, coeficientes)
    
    with open("resultados_python.txt", "w") as f:
        linea80 = "="*80 + "\n"
        linea_sep = "-"*80 + "\n"
        
        f.write(linea80)
        f.write("           RESULTADOS DEL CALCULO POLINOMIAL (Problema 1)\n")
        f.write(linea80)
        f.write(f"CONFIGURACION:\n")
        f.write(f"  Grado (n) : {n}\n")
        f.write(f"  Valor (x) : {x}\n")
        f.write(linea_sep)
        f.write(f"TIEMPO DE EJECUCION: {fin - inicio:.6f} segundos\n")
        f.write(linea_sep + "\n")
        
        f.write(f"RESULTADO FINAL:\n")
        f.write(f"{resultado}\n\n")
        f.write(linea80)

if __name__ == "__main__":
    ejecutar_problema1()

File: Problema1/Solucion.java
import java.io.*;
import java.math.BigInteger;
import java.util.*;

/**
 * UNIVERSIDAD NACIONAL EXPERIMENTAL DE GUAYANA
 * Asignatura: Lenguajes y Compiladores
 * Problema 1: Polinomio (x+1)^n usando Triángulo de Pascal
 */
public class Solucion {
    public static void main(String[] args) throws IOException, InterruptedException {
        Scanner sc = new Scanner(System.in);
        
        while (true) {
            new ProcessBuilder("cmd", "/c", "cls").inheritIO().start().waitFor();
            System.out.println("\n==================================================");
            System.out.println("  EXPANSION POLINOMIAL (TRIANGULO DE PASCAL) - JAVA ");
            System.out.println("==================================================");
            System.out.println("  [1] Usar valores por defecto (n=100, x=2)");
            System.out.println("  [2] Ingresar valores manualmente");
            System.out.println("  [0] Salir");
            System.out.println("--------------------------------------------------");
            System.out.print("Seleccione una opcion > ");
            
            String op = sc.nextLine().trim();
            if (op.equals("0")) {
                System.out.println("Saliendo...");
                break;
            } else if (!op.equals("1") && !op.equals("2")) {
                System.out.println("[!] Opcion no valida.");
                continue;
            }

            int n = 100;
            int x = 2;

            if (op.equals("2")) {
                System.out.println("\n--- CONFIGURACION MANUAL ---");
                while (true) {
                    System.out.print("Ingrese el valor de n: ");
                    String input = sc.nextLine().trim();
                    try {
                        n = Integer.parseInt(input);
                        if (n >= 0) break;
                        System.out.println("Error: n debe ser >= 0.");
                    } catch (NumberFormatException e) {
                        System.out.println("Error: Ingrese un numero entero.");
                    }
                }
                while (true) {
                    System.out.print("Ingrese el valor de x: ");
                    String input = sc.nextLine().trim();
                    try {
                        x = Integer.parseInt(input);
                        break;
                    } catch (NumberFormatException e) {
                        System.out.println("Error: Ingrese un numero entero.");
                    }
                }
            }

            new ProcessBuilder("cmd", "/c", "cls").inheritIO().start().waitFor();
            System.out.printf("\n[INFO] Calculando para n=%d, x=%d...\n", n, x);

            // 1. Generar coeficientes (Fila de Pascal)
            long inicio = System.nanoTime();
            List<BigInteger> fila = new ArrayList<>();
            fila.add(BigInteger.ONE);
            for (int i = 0; i < n; i++) {
                List<BigInteger> nueva = new ArrayList<>();
                nueva.add(BigInteger.ONE);
                for (int j = 0; j < fila.size() - 1; j++) {
                    nueva.add(fila.get(j).add(fila.get(j + 1)));
                }
                nueva.add(BigInteger.ONE);
                fila = nueva;
            }
            double tiempo = (System.nanoTime() - inicio) / 1e9;

            // 2. Calcular y guardar resultados
            try (PrintWriter out = new PrintWriter(new FileWriter("resultados_java.txt"))) {
                String line80 = "================================================================================";
                String lineSep = "--------------------------------------------------------------------------------";
                
                out.println(line80);
                out.println("           RESULTADOS DEL CALCULO POLINOMIAL (Problema 1)");
                out.println(line80);
                out.println("CONFIGURACION:");
                out.printf("  Grado (n) : %d\n", n);
                out.printf("  Valor (x) : %d\n", x);
                out.println(lineSep);
                out.printf("TIEMPO DE EJECUCION: %.6f segundos\n", tiempo);
                out.println(lineSep + "\n");

                BigInteger total = BigInteger.ZERO;
                BigInteger valX = BigInteger.valueOf(x);
                for (int i = 0; i < fila.size(); i++) {
                    int pot = n - i;
                    BigInteger c = fila.get(i);
                    BigInteger termino = c.multiply(valX.pow(pot));
                    total = total.add(termino);
                }

                out.println("RESULTADO FINAL:");
                out.println(total + "\n");
                out.println(line80);
            }
            System.out.printf("[OK] Completado en %.6f s. Revisa 'resultados_java.txt'\n", tiempo);
            System.out.print("\nPresione Enter para continuar...");
            sc.nextLine();
        }
    }
}
